force-app/main/default/classes/Circle.cls

```diff
@@ -0,0 +1,55 @@
+/**
+* @description encapsulates a circle, which is immutable. 
+* It is be possible to construct a circle by providing either it’s radius 
+* or a string representation of its diameter
+*/
+public with sharing class Circle 
+{
+    private Static final Decimal pi = 3.14159; // PI to 5dp
+    public final Decimal radius;
+    public final Decimal circumference;
+    public final Double surfaceArea;
+
+    public Circle(Decimal radius) 
+    {
+        if (radius > 0)
+        {
+            This.radius = radius;
+            This.circumference = This.calculateCircumference();
+            This.surfaceArea = This.calculateSurfaceArea();
+        }
+        else
+        {
+            Throw new CirclePositiveRadiusException('The radius must be > 0');
```

Comment:
This is quite a granular exception - if you went this route in a project then you'd end up with a lot of exceptions that might need to be caught and handled. Again, nothing wrong with it, but it's unusual to have the exception reflect the exact detail. For example, Apex has the NullPointerException, thrown when you attempt to dereference a null value. This doesn't include information about the actual problem, that comes from the message.  If it were me, I'd probably have a CircleException and put the details in the message.